### PR TITLE
CVSL-422:  Change licence history to events, record events and add endpoint to query them

### PR DESCRIPTION
**Pull request by timharrison-moj**
PLEASE DO NOT MERGE - there will be an accompanying UI change to go with this.

* Removes the licence_history table
* Adds a licence_event table
* Records events at all the appropriate places we might be interested in
* Records the variation reason and rejection reasons
* Endpoint to search for events for a licence / all licences / event types.
* Tests and integration tests.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/OffenderService.kt [hunk 1]
```diff
@@ -3,22 +3,40 @@ package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.service
...
     val inFlightLicenceStatuses = listOf(LicenceStatus.IN_PROGRESS, LicenceStatus.SUBMITTED, LicenceStatus.APPROVED, LicenceStatus.ACTIVE)
-
     var offenderLicences = this.licenceRepository.findAllByCrnAndStatusCodeIn(crn, inFlightLicenceStatuses)
 
+    // Update the active licences for this person on probation
     offenderLicences = offenderLicences.map { it.copy(responsibleCom = newCom) }
-
-    // TODO: Create an audit log
-
     this.licenceRepository.saveAllAndFlush(offenderLicences)
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/OffenderService.kt:25]: Implemented this TODO

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt [hunk 2]
```diff
@@ -585,16 +586,18 @@ class LicenceIntegrationTest : IntegrationTestBase() {
       .expectStatus().isOk
 
     val result = webTestClient.get()
-      .uri("/licence/id/1")
+      .uri("/events/match?licenceId=1&eventType=VARIATION_SUBMITTED_REASON")
       .accept(MediaType.APPLICATION_JSON)
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt:590]: Checks that the reason was record in licence_events, not now on the licence itself.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/LicenceEvent.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    example = "Robert Mortimer"
+  )
+  val surname: String? = null,
+
+  @Schema(description = "A reason or description related to the event", example = "Reason for variation")
+  val eventDescription: String? = null,
+
+  @Schema(description = "The date and time of the event", example = "12/01/2022 23:14:23")
+  @JsonFormat(pattern = "dd/MM/yyyy HH:mm:ss")
+  val eventTime: LocalDateTime = LocalDateTime.now(),
```
Comment (smcveigh941) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/LicenceEvent.kt:45]: This should not have a default value. If there is a problem with the database where this value was not populated, we want to know about it by letting it fail, and we shouldnt cover it up by returning an inaccurate value

File: src/test/resources/test_data/clear-all-data.sql [hunk 4]
```diff
@@ -4,7 +4,6 @@ DELETE from additional_condition_upload_detail;
 DELETE from additional_condition_upload_summary;
 DELETE from additional_condition;
 DELETE from bespoke_condition;
-DELETE from licence_history;
```
Comment (smcveigh941) [inline src/test/resources/test_data/clear-all-data.sql:7]: Want to delete the licence events instead?

**Review by smcveigh941 — COMMENTED** [review]
Looks like good work. I believe that eventTime shouldnt have a default value on the DTO though

File: src/test/resources/test_data/clear-all-data.sql (hunk 4 above)
Comment (timharrison-moj) [reply src/test/resources/test_data/clear-all-data.sql:7]: No, its ok, there's a cascading delete for events when deleting the licence.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/LicenceEvent.kt (hunk 3 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/LicenceEvent.kt:45]: Thanks - I'll have a check on this next.

File: src/test/resources/test_data/clear-all-data.sql (hunk 4 above)
Comment (timharrison-moj) [reply src/test/resources/test_data/clear-all-data.sql:7]: Audit events do not cascade deletes - we keep this regardless, but licence events are always linked with a licence and do.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/LicenceEvent.kt (hunk 3 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/LicenceEvent.kt:45]: Yes, you're right. The model object is only used to return licence events from the DB, so if there's no date and time specified there we should return no date and time to the client. Changing this now.

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/model/LicenceEvent.kt]: Updated now.

**Review by smcveigh941 — APPROVED** [review]
LGTM

**Pull request merged**